### PR TITLE
Metadata

### PR DESCRIPTION
**Pull request by Never-M**
1 Create **Write** and **Read** function in gossiper, which allow gossiper write the neighbor information in a csv file before **stop** and read the infomation when creating  a gossiper object in **NewGossiper** function
2 Change all fields of gossiper to public

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -5,23 +5,23 @@ import (
 	gp "github.com/Never-M/MyGossip/pkg/gossiper"
 )
 
-func main()  {
+func main() {
```
Comment (Bowenislandsong) [inline main.go:8]: might wanna just convert this to a e2e test file.

./test/e2e/heart_beat_test.go

File: pkg/gossiper/gossiper.go [hunk 2]
```diff
@@ -18,13 +20,13 @@ const HEARTBEAT_PATH = "heartbeat"
 const HEARTBEAT_TIMEOUT = 1000
 
 type gossiper struct {
```
Comment (Bowenislandsong) [inline pkg/gossiper/gossiper.go]: any reason we want to expose its content but not the struct? 

If you expose the content for retrieving content, then there should be a getter which also dereference these content.

File: pkg/gossiper/gossiper.go [hunk 3]
```diff
@@ -38,53 +40,116 @@ func NewGossiper(name, ip string) *gossiper {
 	logger := Newlogger()
 	logger.SaveToFile("/tmp/" + name + "/log")
 	if err != nil {
-		fmt.Println(err)
 		logger.Fatal("create database Failed", "gossiper", "NewGossiper")
 	}
+	peers := make(map[string]*peer)
+
+	//check if file exsit
+	_, err = os.Stat("/tmp/" + name + "/peers.csv")
```
Comment (Bowenislandsong) [inline pkg/gossiper/gossiper.go]: join path wth filepath.join()

File: pkg/gossiper/gossiper.go [hunk 4]
```diff
@@ -38,53 +40,116 @@ func NewGossiper(name, ip string) *gossiper {
...
+	if err == nil {
```
Comment (Bowenislandsong) [inline pkg/gossiper/gossiper.go:51]: use open or something. This is not a reliable way to see if a file exists.

**Review by Bowenislandsong — COMMENTED** [review]
first quarter review.

File: main.go [hunk 5]
```diff
@@ -5,23 +5,23 @@ import (
...
 	go node2.HeartBeatReceiver()
 	var input string
-	for  {
+	for {
 		fmt.Scanln(&input)
 		if input == "1" {
 			_, err := node1.SendHeartBeats()
 			if err != nil {
-				fmt.Printf("Send req error, %v",err)
+				fmt.Printf("Send req error, %v", err)
```
Comment (wfystx) [inline main.go]: use log

File: pkg/gossiper/gossiper.go [hunk 6]
```diff
@@ -38,53 +40,116 @@ func NewGossiper(name, ip string) *gossiper {
...
+		//exsit!
+		peerpairslice := Read(name)
+		if len(peerpairslice) > 0 {
+			for _, item := range peerpairslice {
```
Comment (wfystx) [inline pkg/gossiper/gossiper.go]: peerPairSlice

File: pkg/gossiper/gossiper.go [hunk 7]
```diff
@@ -38,53 +40,116 @@ func NewGossiper(name, ip string) *gossiper {
...
+	}
+}
+
+type PeerPair struct {
+	Name string
+	IP   string
+}
+
+func (g *gossiper) Write() {
+	var tosave []PeerPair
```
Comment (wfystx) [inline pkg/gossiper/gossiper.go]: toSave

File: pkg/gossiper/gossiper.go [hunk 8]
```diff
@@ -38,53 +40,116 @@ func NewGossiper(name, ip string) *gossiper {
...
+		Logger:        logger,
+	}
+}
+
+type PeerPair struct {
+	Name string
+	IP   string
+}
+
+func (g *gossiper) Write() {
```
Comment (wfystx) [inline pkg/gossiper/gossiper.go]: A clearer name would be better. eg. WritePeersToFile

File: pkg/gossiper/gossiper.go [hunk 9]
```diff
@@ -38,53 +40,116 @@ func NewGossiper(name, ip string) *gossiper {
...
+		line := []string{pair.Name, pair.IP}
+		err = writer.Write(line)
+		if err != nil {
+			g.Logger.Panic("Write Failed -- <err>: ")
+		}
+	}
+	writer.Flush()
+}
+
+func Read(name string) []PeerPair {
```
Comment (wfystx) [inline pkg/gossiper/gossiper.go]: Maybe ReadPeersFromFile

File: pkg/gossiper/gossiper.go [hunk 10]
```diff
@@ -38,53 +40,116 @@ func NewGossiper(name, ip string) *gossiper {
...
+	}
+	defer file.Close()
+
+	reader := csv.NewReader(file)
+	reader.FieldsPerRecord = -1
+	record, err := reader.ReadAll()
+	if err != nil {
+		logger.Panic("Failed to read csv")
+	}
+	var peerpairslice []PeerPair
```
Comment (wfystx) [inline pkg/gossiper/gossiper.go]: peerPairSlice

File: main.go (hunk 1 above)
Comment (Never-M) [reply main.go:8]: I discussed this with Fuyao. He said he didn't have enough time to do this. So maybe ignore this for now?

**Review by wfystx — APPROVED** [review]
Good work

File: pkg/gossiper/gossiper.go [hunk 11]
```diff
@@ -36,7 +36,7 @@ type heartBeat struct {
 	Time string `json:"time"`
 }
 
-func NewGossiper(name, ip string) *gossiper {
+func NewGossiper(name, ip string) *Gossiper {
 	_, db, err := Newdb(filepath.Join("/tmp/", name, "/database"))
```
Comment (Bowenislandsong) [inline pkg/gossiper/gossiper.go:40]: any reason why this has to be in a tmp dir? we could have this as an input parameter and let it go to tmp is input is "".

**Pull request merged**